src/detectors/transpilers/html/parser.ts

```diff
@@ -0,0 +1,73 @@
+import type {ReadStream} from "node:fs";
+import {Detail, SaxEventType, SAXParser, Tag as SaxTag} from "sax-wasm";
+import {finished} from "node:stream/promises";
+import fs from "node:fs/promises";
+import {createRequire} from "node:module";
+const require = createRequire(import.meta.url);
+
+let saxWasmBuffer: Buffer;
+async function initSaxWasm() {
+	if (!saxWasmBuffer) {
+		const saxPath = require.resolve("sax-wasm/lib/sax-wasm.wasm");
+		saxWasmBuffer = await fs.readFile(saxPath);
+	}
+
+	return saxWasmBuffer;
+}
+
+async function parseHtml(contentStream: ReadStream, parseHandler: (type: SaxEventType, tag: Detail) => void) {
+	const options = {highWaterMark: 32 * 1024}; // 32k chunks
+	const saxWasmBuffer = await initSaxWasm();
+	const saxParser = new SAXParser(SaxEventType.CloseTag, options);
+
+	saxParser.eventHandler = parseHandler;
+
+	// Instantiate and prepare the wasm for parsing
+	if (!await saxParser.prepareWasm(saxWasmBuffer)) {
+		throw new Error("Unknown error during WASM Initialization");
+	}
+
+	// stream from a file in the current directory
+	contentStream.on("data", (chunk: Uint8Array) => {
+		try {
+			saxParser.write(chunk);
+		} catch (err) {
+			if (err instanceof Error) {
+				// In case of an error, destroy the content stream to make the
+				// error bubble up to our callers
+				contentStream.destroy(err);
+			} else {
+				throw err;
+			}
+		}
+	});
+	await finished(contentStream);
+	saxParser.end();
+}
+
+export async function extractJSScriptTags(contentStream: ReadStream) {
+	const scriptTags: SaxTag[] = [];
+
+	await parseHtml(contentStream, (event, tag) => {
+		if (tag instanceof SaxTag &&
+			event === SaxEventType.CloseTag &&
+			tag.value === "script") {
+			const isJSScriptTag = tag.attributes.every((attr) => {
+				// The "type" attribute of the script tag should be
+				// 1. not set (default),
+				// 2. an empty string,
+				// 3. or a JavaScript MIME type (text/javascript)
+				// https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type#attribute_is_not_set_default_an_empty_string_or_a_javascript_mime_type
+				return attr.name.value !== "type" ||
+					(attr.name.value === "type" &&
+					(attr.value.value === "" || attr.value.value === "text/javascript"));
```

Comment:
In the scope of SAPUI5, I find ~180 occurrences of the legacy MIME type `application/javascript`. Maybe we should include that one as well? The other legacy types, I didn't find.

According to https://mimesniff.spec.whatwg.org/#javascript-mime-type-essence-match, the MIME types are case insensitive. I don't expect the SAX parser to normalize attribute values, so the check maybe should be case insensitive.

Last but not least: we most likely won't process scripts of type `module` any time soon, but the check for inline scripts could complain about them already (if they're inline)

Reply:
Thanks, @codeworrior !

I have added the case insensitive checks and also the `application/javascript` type.

Regarding the `module` type, currently it's being ignored as we check only for missing `type` property or against a list of hardcoded, case insensitive (already, thanks!) types.

I have added type="module" to the test cases

Reply:
@codeworrior's comment wasn't fully addressed, right? I also agree that the check should complain about type="module", as it is also relevant for CSP unsafe-inline.

In addition, a <script> tag with a "src" attribute and a comment (or even code) as content is falsely reported as finding. Browsers (tested with Chrome only) ignore the script tag content in case a "src" attribute is provided (even without a value).
```html
<script src="foo.js"> // should not be reported as it is not a CSP violation
</script>
```
```html
<script src> // should also not be reported as it is not a CSP violation
console.log("this code won't run");
</script>
```